### DISCUSSION TITLE
Is there a way to build_tx without incrementing the internal address index?

### DISCUSSION THREAD

**Discussion by ConorOkus** [bitcoindevkit/bdk#948] [Q&A]
Repost from Discord by @icota.

**Comment by ConorOkus** [comment]
@thunderbiscuit:

"One way you can do that is by using one of the more fine-tuning variants on the AddressIndex type when calling the Wallet.get_address() method. For example you can use AddressIndex.Peek(u32) always with the same index number, or AddressIndex.Reset(u32) to keep the index at a specific number."

**Comment by ConorOkus** [comment]
@notmandatory:

"In addition to what @thunderbiscuit said above you should also be able to use the .drain_to() function to use your self selected address for transaction change: https://docs.rs/bdk/latest/bdk/wallet/tx_builder/struct.TxBuilder.html#method.drain_to"

**Comment by LLFourn** [comment]
In `bdk v1.0.0-alpha.0` we added the ability to "cancel" a tx after it has been built which releases the change address so the next time you build a transaction you get it back. That might be what the user was after here. I can't link to the docs because they're broken #955.

**Comment by icota** [comment]
@LLFourn @notmandatory @thunderbiscuit  would having a `AddressIndex::Current` and `change_address_index` to pass it onto the builder be an acceptable solution to this? Like here: https://github.com/icota/bdk/commit/ca050642692fa0aaee6bf347ffb452c086821510

I think it's more elegant than cancelling the transaction every time? In this approach (BDK) user increments the change when transaction is broadcast or PSBT is shown to end user

**Reply by thunderbiscuit** [reply]
I think this makes sense. I agree we do need more fine control over how the wallet increments the index.

**Reply by LLFourn** [reply]
I think we have already made passing the change address explicitly possible but you also probably want a way for BDK to manage this in a sane and thread-safe manner too (two threads creating txs at the same time get different change addresses).

> I think it's more elegant than cancelling the transaction every time? In this approach (BDK) user increments the change when transaction is broadcast or PSBT is shown to end user

You wouldn't cancel it every time but rather cancel it when you don't intend to broadcast the resulting transaction. This seems like a necessary party of the problem if you are using a derivation index based scheme. If you no longer intend to use the change address index you have to unreserve it somehow. It seems like in the commit you referenced you do allow for re-use. Sure if you don't care too much about re-use or have some other mechanism to prevent it then there's no need to use the cancel tx method system.

**Reply by icota** [reply]
> I think we have already made passing the change address explicitly possible

@LLFourn but is there a way to get the address at current index regardless of whether it's been previously used? I couldn't find that

**Reply by LLFourn** [reply]
In `1.0.0-alpha.0` there is`Wallet::derivation_index(keychain)` that you can use to get the index and then use the `Peek` index to get it. In `SpkTxOutIndex` you can use `spk_at_index`.
